src/solvers/refinement/string_constraint_generator_valueof.cpp

```diff
@@ -149,47 +187,44 @@ string_exprt string_constraint_generatort::add_axioms_from_int(
   for(size_t size=1; size<=max_size; size++)
   {
     // For each possible size, we have:
-    // sum_0 = starts_with_digit ? res[0]-'0' : 0
-    // sum_j = 10 * sum_{j-1} + res[i] - '0'
+    // sum_0 = numeric value of res[0] (which is 0 if res[0] is '-')
+    // sum_j = radix * sum_{j-1} + numeric value of res[j]
     // and add axioms:
     // a5 : |res| == size =>
     //        forall 1 <= j < size.
     //          is_number && (j >= max_size-2 => no_overflow)
-    //     where is_number := '0' <= res[j] <= '9'
-    //     and no_overflow := sum_{j-1} = (10 * sum_{j-1} / 10)
+    //     where is_number := res[j] is a digit for the radix
+    //     and no_overflow := sum_{j-1} = (radix * sum_{j-1} / radix)
     //                        && sum_j >= sum_{j - 1}
     //         (the first part avoid overflows in multiplication and
     //          the second one in additions)
-    // a6 : |res| == size && '0' <= res[0] <= '9' => i == sum
-    // a7 : |res| == size && res[0] == '-' => i == -sum
+    // a6 : |res| == size && res[0] is a digit for radix => x == sum
+    // a7 : |res| == size && res[0] == '-' => x == -sum
 
     exprt::operandst digit_constraints;
-    exprt sum=if_exprt(
-      starts_with_digit,
-      typecast_exprt(minus_exprt(res[0], zero_char), type),
-      from_integer(0, type));
+    exprt sum=
+      get_numeric_value_from_character(res[0], char_type, type, radix_ul);
```

Comment:
What happens if `radix` is not a constant? This will probably not be correct, this will probably almost never happen, but it would be nice to document this somewhere and maybe have a regression/unit test for that.